File: lua/langs/rust.lua
vim.pack.add({ {
	src = "https://github.com/mrcjkb/rustaceanvim",
	version = vim.version.range("^9"),
} })

vim.g.rustaceanvim = {
	server = {
		default_settings = {
			["rust-analyzer"] = {
				assist = {
					preferSelf = true,
				},
			},
		},
	},
}

vim.api.nvim_create_autocmd("FileType", {
	pattern = "rust",
	callback = function()
		vim.o.shiftwidth = 0
		vim.o.tabstop = 4
	end,
})

vim.api.nvim_create_autocmd("FileType", {
	pattern = "rust",
	callback = function()
		require("nvim-treesitter").install({ "rust" })
	end,
	once = true,
})

require("conform").formatters_by_ft.rust = { "rustfmt" }

File: src/langs/rust.rs
use crate::{env::gvim::pack::PackOpts, prelude::*};

impl NvimConf<'_> {
    pub fn load_rust_lang(&self) {
        self.ft_set_indent("rust", 4);
        self.set_formatter("rust", ["rustfmt"]);

        if self.is_vscode() {
            return;
        }

        if let Some(version) = self.version_range("^9").ok_or_notify(self) {
            self.add_packs([mk_builder!(PackOpts, {
                src = "https://github.com/mrcjkb/rustaceanvim";
                version = version;
            })]);
        } else {
            return;
        }

        self.with_vim_g(|g| {
            tbl!(out(g), {
                rustaceanvim = self.rustaceanvim_opts();
            })
        })
        .ok_or_notify(self);
    }

    fn rustaceanvim_opts(&self) -> impl LuaSub<LuaVal> {
        let ra_opts = tbl!(owned, {
            assist = tbl!(owned, {
                preferSelf = true;
                // cargo = tbl!(owned, {
                //     target = "";
                // });
            });
        });

        tbl!(owned, {
            server.default_settings = tbl!(owned, {
                "rust-analyzer" = ra_opts;
            });
        })
    }
}
